Add MCS ResourceImport controller

abhiraut opened the pull request.

abhiraut committed:
Add MCS ResourceImport controller

luolanzone commented on multicluster/controllers/multicluster/resourceimport_controller.go:
```diff
@@ -18,42 +18,210 @@ package multicluster
+		// TODO: Update internal cache
+		err := (*r.localClusterManager).Create(ctx, mcsObj, &client.CreateOptions{})
+		if err != nil {
+			klog.Error("fail to create ResourceImport %s/%s,err: %v", mcsObj.GetNamespace(), mcsObj.GetName(), err)
```
s/ResourceImport/Service/

luolanzone replied:
done in a new commit

luolanzone commented on multicluster/cmd/multicluster-controller/controller.go:
```diff
@@ -168,12 +168,15 @@ func run(o *Options) error {
+		mgr.GetClient(),
+		mgr.GetScheme(),
+		localMgr,
+		remoteMgr,
```
I think common area doesn't support initialize both localMgr and remoteMgr now,  it will only create remoteMgr when it's member cluster and create localMgr when it's leader cluster. need @aravindakidambi to double check and clarify this.

luolanzone replied:
I have refined `clusterset_controller.go` so it can support to run as both leader and member in a new commit

luolanzone commented on multicluster/controllers/multicluster/resourceimport_controller.go:
```diff
@@ -18,42 +18,210 @@ package multicluster
+		installedSvcs: cache.NewIndexer(svcInfoKeyFunc, cache.Indexers{
+			svcIndexerByType: svcIndexerByTypeFunc,
+		}),
+		installedResImports: cache.NewIndexer(resImportIndexerKeyFunc, cache.Indexers{
```
looks like this `installedResImports` is created but not used.

luolanzone committed:
update importer and fix issue

1. add Endpoint create/update, Service's update logic
2. add some missing roles
3. support a cluster to run as both leader and member
4. add a new config `webhookServer` in config CRD to skip webhook installation
so we can support instance only installation

Signed-off-by: Lan Luo <luola@vmware.com>

luolanzone commented:
importer related codes have been included in #29 , so close this one.
